filechanges/opt/musicbox/read_ini.sh

```diff
@@ -11,281 +11,277 @@
-    local INI_ALL_SECTION="${VARNAME_PREFIX}__ALL_SECTIONS"
-    local INI_NUMSECTIONS_VARNAME="${VARNAME_PREFIX}__NUMSECTIONS"
-    if [ "${CLEAN_ENV}" = 1 ] ;then
-	eval unset "\$${INI_ALL_VARNAME}"
-    fi
-    unset ${INI_ALL_VARNAME}
-    unset ${INI_ALL_SECTION}
-    unset ${INI_NUMSECTIONS_VARNAME}
-
-    if [ -z "$INI_FILE" ] ;then
-	cleanup_bash
-	return 0
-    fi
-    
-    if ! check_ini_file ;then
-	cleanup_bash
-	return 1
-    fi
-
-    # Sanitise BOOLEANS - interpret "0" as 0, anything else as 1
-    if [ "$BOOLEANS" != "0" ]
-    then
-	BOOLEANS=1
-    fi
-
-
-    # }}} END Options
-
-    # }}} END Deal with command line args
-
-    local LINE_NUM=0
-    local SECTIONS_NUM=0
-    local SECTION=""
-    
-    # IFS is used in "read" and we want to switch it within the loop
-    local IFS=$' \t\n'
-    local IFS_OLD="${IFS}"
-    
-    # we need some optional shell behavior (shopt) but want to restore
-    # current settings before returning
-    local SWITCH_SHOPT=""
-    pollute_bash
-    
-    while read -r line || [ -n "$line" ]
-    do
+	local LINE_NUM=0
+	local SECTIONS_NUM=0
+	local SECTION=""
+	
+	# IFS is used in "read" and we want to switch it within the loop
+	local IFS=$' \t\n'
+	local IFS_OLD="${IFS}"
+	
+	# we need some optional shell behavior (shopt) but want to restore
+	# current settings before returning
+	local SWITCH_SHOPT=""
+	pollute_bash
+	
+	while read -r line || [ -n "$line" ]
+	do
 #echo line = "$line"
 
-	((LINE_NUM++))
+		((LINE_NUM++))
 
-	# Skip blank lines and comments
-	if [ -z "$line" -o "${line:0:1}" = ";" -o "${line:0:1}" = "#" ]
-	then
-	    continue
-	fi
+		# Skip blank lines and comments
+		if [ -z "$line" -o "${line:0:1}" = ";" -o "${line:0:1}" = "#" ]
+		then
+			continue
+		fi
 
-	# Section marker?
-	if [[ "${line}" =~ ^\[[a-zA-Z0-9_]{1,}\]$ ]]
-	then
+		# Section marker?
+		if [[ "${line}" =~ ^\[.*\]$ ]]
+		then
+			# Set SECTION var to name of section (strip [ and ] from section marker)
+			SECTION="${line#[}"
+			SECTION="${SECTION%]}"
+			read SECTION <<<"${SECTION}"
+			if ! [[ "${line}" =~ ^[[:print:]]+$  ]] ;then
+				echo "Error: Invalid section:" >&2
+				echo " ${LINE_NUM}: '$line'" >&2
+				cleanup_bash
+				return 1
+			fi
+			((SECTIONS_NUM++))
+
+			continue
+		fi
 
-	    # Set SECTION var to name of section (strip [ and ] from section marker)
-	    SECTION="${line#[}"
-	    SECTION="${SECTION%]}"
-	    eval "${INI_ALL_SECTION}=\"\${${INI_ALL_SECTION}# } $SECTION\""
-	    ((SECTIONS_NUM++))
+		# Are we getting only a specific section? And are we currently in it?
+		if [ ! -z "$INI_SECTION" ]
+		then
+			if [ "$SECTION" != "$INI_SECTION" ]
+			then
+				continue
+			fi
+		fi
 
-	    continue
-	fi
+		# Valid var/value line? (check for variable name and then '=')
+		# HACK: Handling of local-sqlite's 'directories' variables. Bypass lines ending in 'directory:'
+		#       as if they are valid.
+		if ! [[ "${line}" =~ ^[[:print:]]+[[:space:]]*directory:|= ]]
```

Comment:
There's a comment in https://github.com/rudimeier/bash_ini_parser/issues/2 explaining the specific scenario that this hack is meant to work around. The local-sqlite 'directories' parameter is unusually complex in that it consists of multiple lines of values, each potentially also containing an '=' sign:

```
# top-level directories for browsing, as <name> <uri>
directories =
    Albums                  local:directory?type=album
    Artists                 local:directory?type=artist
    Composers               local:directory?type=artist&role=composer
    Folders                 local:directory:
    Genres                  local:directory?type=genre
    Performers              local:directory?type=artist&role=performer
    Release Years           local:directory?type=date&format=%25Y
    Tracks                  local:directory?type=track
    Last Week's Updates     local:directory?max-age=604800
    Last Month's Updates    local:directory?max-age=2592000
```

I'm not sure how to process something like that in bash and still get an array of valid key/value pairs. So for now, we just want the script to run through as we won't use these specific variables in any of the Pi Musicbox scripts. The hack in the regular expression is just supposed to handle this line:

```Folders                 local:directory:```

...because the line does not contain an '=' sign it is not seen as a valid key/value pair and the script fails. By adding ```directory:``` to the regular expression, lines ending in 'directory:` will be parsed as well just to handle this one exception.